_posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd

```diff
@@ -0,0 +1,390 @@
+
+## Terminological introduction
+
+"The JIT", much talked about in PyTorch-world and an eminent feature of R `torch`, as well, is two things at the same time -- depending on how you look at it: an optimizing compiler; and a free pass to execution in many environments where neither R nor Python are present.
+
+#### Compiled, interpreted, just-in-time compiled
+
+Starting with the former: "JIT" is a common acronym for "just in time" [to wit: compilation]. *Compilation* means generating machine-executable code; it is something that has to happen to every program for it to be runnable. The question is when.
+
+C code, for example, is compiled "by hand", at some arbitrary time prior to execution. Many other languages, however (among them Java, R, and Python) are -- in their default implementations, at least -- *interpreted*: They come with executables (`java`, `R`, and `python`, resp.) that create machine code *at run time*, based on either the original program as written or an intermediate format called *bytecode*. Interpretation can proceed line-by-line, such as when you enter some code in R's REPL (read-eval-print loop), or in chunks (if there's a whole script or application to be executed). In the latter case, since the interpreter knows what is likely to be run next, it can implement optimizations that would be impossible otherwise. This process is commonly known as *just-in-time compilation*. Thus, in general parlance, JIT compilation is compilation, but at a point in time where the program is already running.
+
+#### The `torch` just-in-time compiler
+
+Compared to that notion of JIT, at once generic (in technical regard) and specific (in time), what (Py-)Torch people have in mind when they talk of "the JIT" is both more narrowly-defined (in terms of operations) and more inclusive (in time): What is understood is the complete process from providing code input that can be converted into an intermediate representation (IR), via generation of that IR, via successive optimization of the same by the JIT compiler, via conversion (again, by the compiler) to bytecode, to -- finally -- execution, again taken care of by that same compiler, that now is acting as a virtual machine.
+
+If that sounded complicated, don't be scared. To actually make use of this feature from R, not much is to be learned in terms of syntax; a single function, augmented by a few specialized helpers, is stemming all the heavy load. What matters though is to understand a bit about how JIT compilation works, so you know what to expect, and are not surprised by unintended outcomes.
+
+## What's coming (in this text)
+
+This post has three further parts.
+
+In the first, we explain how to make use of JIT capabilities in R `torch`. Beyond the syntax, we focus on the semantics: what essentially happens when you "JIT trace" a piece of code, and how that affects the outcome.
```

Comment:
```suggestion
In the first, we explain how to make use of JIT capabilities in R `torch`. Beyond the syntax, we focus on the semantics (what essentially happens when you "JIT trace" a piece of code), and how that affects the outcome.
```

Reply:
haha the usual (colon ;-))
guess i need to not be stubborn at least once in a while :-))